imports/lib/models/Model.ts

```diff
@@ -0,0 +1,187 @@
+  [K in keyof T]: undefined extends z.input<T[K]> ? T[K] : z.ZodOptional<T[K]>;
+}, UnknownKeys, Catchall> {
+  const newShape: any = {};
+  for (const [key, fieldSchema] of Object.entries(schema.shape)) {
+    newShape[key] = fieldSchema.isOptional() ? fieldSchema : fieldSchema.optional();
+  }
+
+  return new z.ZodObject({
+    ...schema._def,
+    shape: () => newShape,
+  });
+}
+
+const modifierIsWholeDoc = <T extends Document>(modifier: Mongo.Modifier<T>): modifier is T => {
+  return !Object.keys(modifier).some((k) => k.startsWith('$'));
+};
+
+export const AllModels = new Set<Model<any>>();
+
+class Model<T extends z.ZodObject<any, 'strict'>> {
+  name: string;
+
+  schema: T;
+
+  collection: Mongo.Collection<z.output<T>>;
+
+  constructor(name: string, schema: T) {
+    this.name = name;
+    this.schema = schema;
+    this.collection = new Mongo.Collection(name);
+    AllModels.add(this);
+  }
+
+  async insertAsync(doc: z.input<T>): Promise<z.output<T>['_id']> {
+    const cleaned: z.output<T> = await IsInsert.withValue(true, async () => {
+      return this.schema.parseAsync(doc);
+    });
+    return this.collection.insert(cleaned as any) as any;
+  }
+
+  // As a note, we don't accept full documents as modifiers because our
+  // transform functions don't make sense (e.g. we won't accept a `createdAt`
+  // timestamp, but we expect one to be there)
+  async updateAsync(
+    selector: Selector<z.output<T>>,
+    modifier: Exclude<Mongo.Modifier<z.output<T>>, z.output<T>>,
+    options: {
+      multi?: boolean | undefined;
+      upsert?: boolean | undefined;
+      arrayFilters?: { [identifier: string]: any }[] | undefined;
+    } = {},
+  ): Promise<number> {
+    // Types should prevent passing a full document as a modifier, but verify at
+    // runtime to be sure
+    const rawModifier: Mongo.Modifier<z.output<T>> = modifier as any;
+    if (modifierIsWholeDoc(rawModifier)) {
+      throw new Error('Cannot update with a full document');
+    }
+
+    const setOnInsertFields = await IsUpsert.withValue(true, async () => {
+      return optionalPreservingPartial(this.schema).parseAsync({});
+    });
+    const setFields = await IsUpdate.withValue(true, async () => {
+      return optionalPreservingPartial(this.schema).parseAsync({});
+    });
+
+    const mergedModifier: Mongo.Modifier<z.output<T>> = {
+      ...rawModifier,
+      $setOnInsert: {
+        ...rawModifier.$setOnInsert,
+        ...setOnInsertFields,
+      },
+      $set: {
+        ...rawModifier.$set,
+        ...setFields,
+      },
+    };
+
+    return this.collection.update(selector, mergedModifier, options);
+  }
+
+  async upsertAsync(
+    selector: Selector<z.output<T>>,
+    modifier: Exclude<Mongo.Modifier<z.output<T>>, z.output<T>>,
+    options: {
+      multi?: boolean | undefined;
+    } = {},
+  ): Promise<{
+    numberAffected?: number | undefined;
+    insertedId?: z.output<T>['_id'] | undefined;
+  }> {
```

Comment:
This  implementation seems to result in an `updatedAt` field (using `updatedTimestamp`) being set on upserts-that-are-inserts, which is a behavioral difference from what we had before (which was that `createdAt` would be set on first insert, and `updatedAt` would only be set on subsequent updates).

Reply:
Yeah, OK - I think I can get the right behavior if I (a) treat `IsUpsert` and `IsUpdate` as mutually exclusive (which matches the behavior of collection2) (b) if fields are set in both `$set` and `$setOnInsert`, drop them from `$setOnInsert`.

(I'll note that the `updatedAt` timestamp on `Servers` behaves differently than other `updatedAt` timestamps today - it _is_ written on insert as well as update)